File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.ts
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */
    topKFrequent(nums: number[], k: number): number[] {
        const freqMap = new Map<number, number>;
        for (const num of nums) {
            if (!freqMap.has(num)) {
                freqMap.set(num, 1);
                continue;
            }
            freqMap.set(num, freqMap.get(num)+1);
        }

        const groups: number[][] = new Array().fill([]);
        for (let i = 0; i <= nums.length; i++) {
            const entries = Array.from(freqMap.entries(), entry => { const[num, freq] = entry; return freq === i ? entry : null; }).filter(Boolean);
            const nums = entries.map(tuple => tuple[0]);
            groups[i] = nums;
        }

        const result: number[] = [];
        for (let i = nums.length; i >= 0; i--) {
            for (const n of groups[i]) {
                result.push(n);
                if (result.length === k) {
                    return result;
                }
            }
        }
        
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefixProds = [1]*len(nums)
        for i in range(1, len(nums)):
            prefixProds[i] = prefixProds[i-1]*nums[i-1]

        suffixProds = [1]*len(nums)
        for i in range(len(nums)-1, 0, -1):
            suffixProds[i-1] = suffixProds[i]*nums[i]

        result = []
        for i in range(len(nums)):
            result.append(prefixProds[i]*suffixProds[i])
        
        return result
        